fix/layout bugs in help modal

saubhagya-patel opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #30781

This PR solves the layout bugs in the help modal.

### Tasks
- [x] The thumb-up emoji is too low on “React to selected message with 👍”.
- [x] The Time example output sits too high in its table row.

### Changes Made
- For the thumbs-up emoji fix, add `margin-bottom: 7px;` to the `image` element having the `.emoji` class.
- For the time example output fix, add `margin-top: var(--markdown-interelement-space-px);`  to `<p>` elements containing a `<span>` with the `.timestamp-content-wrapper` class.


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

| Before Changes | After Changes |
| -------------- | ------------- |
| ![Before 1](https://github.com/user-attachments/assets/66e8e336-4347-4a76-9604-374959061e33) | ![After 1](https://github.com/user-attachments/assets/34f9ff3a-7d36-4d96-9a3d-56fb86fa4401) |
| ![Before 2](https://github.com/user-attachments/assets/a270660f-226e-428e-8498-680f4fb11633) | ![After 2](https://github.com/user-attachments/assets/e10d3c65-ba80-4616-b5ae-90477fff4485) |

### Reflected Changes in the help modal
- Compact Mode: Disabled
<details><summary>For the thumbs-up emoji fix.</summary>
<p>

![image](https://github.com/user-attachments/assets/b3490277-d1af-4e5e-886f-021430050388)

</p>
</details> 

<details><summary>For the time example output fix.</summary>
<p>

![image](https://github.com/user-attachments/assets/8a6a7efa-d71f-487c-8579-af4f3d9a5105)


</p>
</details> 

- Compact Mode: Enabled

<details><summary>For the thumbs-up emoji fix.</summary>
<p>

![image](https://github.com/user-attachments/assets/00747981-7652-4d1b-ac09-d2ec161453f9)


</p>
</details> 

<details><summary>For the time example output fix.</summary>
<p>

![image](https://github.com/user-attachments/assets/f5a42763-50e1-4dc1-b276-cbd23d48a977)


</p>
</details> 


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks for the fix! Could you please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)?

alya commented:
Also, the screenshots in the PR description should have some surrounding context, to make it possible to tell what looks good.

saubhagya-patel commented:
Thank you for your valuable feedback and for reviewing the PR. I have made the proposed changes to the PR and updated the commit messages to align with the commit style guidelines.
I apologize for any inconvenience caused and look forward to your review of the updated changes.

alya commented:
Thanks! Please add screenshots with "[compact mode](https://zulip.com/help/font-size)" both enabled and disabled (I'm not sure which one you have so far).

saubhagya-patel commented:
Thank you for the ongoing feedback. I have added the screenshots with the "compact mode" both enabled and disabled in the PR description. If there are any additional adjustments needed, please feel free to let me know, and I will update my PR accordingly.

alya commented:
Thanks! @karlstolley are you up for reviewing this one? I haven't tested.

karlstolley commented on web/styles/rendered_markdown.css:
```diff
@@ -10,6 +10,10 @@
         margin-top: var(--markdown-interelement-doubled-space-px);
     }
 
+    & p:has(span.timestamp-content-wrapper) {
```
This is incorrect; this affects all rendered markdown. There needs to be a specific rule in the `informational_overlays.css` file for this.

But beyond that, I think removing the bottom margin (`--markdown-interelement-space`) on ALL elements in the informational overlay produces a much better output for all elements. In that case, the padding on the given `<td>` is in charge of spacing.

saubhagya-patel replied:
I apologize for any inconvenience caused.
To address the issue, adding the following rule to the `informational_overlays.css` file will resolve the problem:
```
& p>time>span.timestamp-content-wrapper {
    margin-top: var(--markdown-interelement-space-px);
}
```
### Screenshots
| Before Changes | After Changes |
|----------|----------|
| ![image](https://github.com/user-attachments/assets/2d770245-4b31-4ef5-94a5-8cbeda66249d)| ![image](https://github.com/user-attachments/assets/23b3f6e3-3e05-40ff-a251-33a63927fa8a)|


However, once I hear your thoughts on this, I think I will proceed to open a discussion in CZO. Thank you.

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -65,6 +65,10 @@
         text-align: right;
+
+        & img.emoji {
+            margin-bottom: 7px;
```
Let's not do this; study how inline emoji are presented in all rendered markdown areas. The "React to selected message with 👍🏻" needs to take the same stack of CSS variables, and then it should be perfectly aligned with the surrounding text.

saubhagya-patel replied:
Thanks for the valuable feedback.
On studying the inline emoji I found that the above problem can indeed be solved using the following modification:
```
& .emoji {
            margin: var(--length-line-oversize-block-margin-adjust) auto;
            vertical-align: sub;
        }
```

### Screenshots

| Before Changes | After Changes |
|----------|----------|
| ![image](https://github.com/user-attachments/assets/a0c23180-9a63-4620-8cfb-30800e95776d)| ![image](https://github.com/user-attachments/assets/c601e915-2cef-451c-a197-e43b5a5a3a60)|

I would like to hear what you think about this before moving forward.

karlstolley replied:
The styles here should match the `:heart:` emoji on the Message formatting pane, specifically:

![Screenshot 2024-10-14 at 2 47 41 PM](https://github.com/user-attachments/assets/24af399f-8ed7-4473-87e4-1bab51d09a64)

There's no need to invent hacky workarounds like `vertical-align: sub`

saubhagya-patel replied:
I tried the approach mentioned above and got the following result:

<details> 
<summary> Screenshot of thumbs-up emoji</summary>

![image](https://github.com/user-attachments/assets/766c1cd9-463a-4bb7-8761-498dd7bb81e2)

</details>

On examining the emoji, I think it is due to the shape of the thumbs-up emoji specifically that it looks like it is misaligned if we replace the thumbs-up emoji with a heart emoji, it aligns well. (Screenshot follows)

<details> 
<summary> Screenshot of heart emoji</summary>

![image](https://github.com/user-attachments/assets/22403853-a3df-467e-b3fc-cd631ff3dede)

</details>

That's why I figured to use the `vertical-align: sub` property.

karlstolley commented:
Thank you for your contribution, @saubhagya-patel! I've left some direct notes in your code; regarding eliminating `margin-bottom` on all the rendered markdown in the overlay, we might need to open a discussion in CZO in the `#design` channel to see if that additional space is worth declaring some other way (e.g., by adding to the padding on the `<td>`) or if it just looks better without the margin entirely--allowing for more content on screen (which I'm in favor of).

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -103,6 +108,10 @@
     margin-bottom: 10px !important;
 }
 
+& p > time > span.timestamp-content-wrapper {
```
There must be a simpler selector than this to get the needed change here. `.timestamp-content-wrapper` is plenty specific; it just needs to be isolated to the message-formatting table somehow.

saubhagya-patel replied:
Thank you for your review. I’ve updated the CSS selector as follows:
```
#message-formatting p:has(time) {
    margin-top: var(--markdown-interelement-space-px);
}
```
This change should address the feedback and achieve the desired result.
I look forward to your review of the updated changes.

karlstolley replied:
That's an improvement, for sure. I still think it would be good to open a #design discussion about removing inter-element spacing on all of the modal examples, and then this brittle selector would become unnecessary entirely.

saubhagya-patel replied:
Thank you for your review. In response, I've opened a [discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/layout.20bugs.20in.20the.20help.20modal) in the design channel and would like to hear your feedback on this topic.

saubhagya-patel commented:
Based on the [CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/layout.20bugs.20in.20the.20help.20modal)  regarding the removal of the bottom margin from block-level rendered markdown elements in the Message Formatting section of the help modal, I have updated this PR accordingly.

I have introduced a new class named `preserve-message-formatting-layout` that sets bottom-margin to zero:
```
.preserve-message-formatting-layout {
    margin-bottom: unset !important;
}
```

To dynamically add the above class, I utilized the `$markdown_help` jQuery object from the `info_overlay.ts` file. I applied the class to `p`, `ul`, `ol`, `pre`, and `blockquote` elements within the `$markdown_help` object using the `process_message_formatting_layout` function.

I used the `!important `property to ensure the style is applied, as it was being overridden by the following selector from
 `rendered_markdown.css` file:
```
.rendered_markdown p {
    margin: 0 0 var(--markdown-interelement-space-px);
}
```

Thank you.

karlstolley commented on web/src/info_overlay.ts:
```diff
@@ -359,3 +360,11 @@ export function show(target: string | undefined): void {
 }
+
+function process_message_formatting_layout($rendered_html: JQuery): void {
+    $rendered_html.find("p").addClass("preserve-message-formatting-layout");
```
I think you can pass a selector like "p, ul, ol, pre, blockquote" to `.find()`, though if these are all immediate children, `.children()` would be the better jQuery function, I think.

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -126,3 +126,7 @@
 }
+
+.preserve-message-formatting-layout {
+    margin-bottom: unset !important;
```
Is it possible to double-up on a class selector here rather than resorting to `!important`? Presumably `.rendered_markdown p` is what you're up against?

saubhagya-patel replied:
Yes, you are right. The previous selector was overridden by the following selector from `rendered_markdown.css` file:
```
.rendered_markdown p {
    margin: 0 0 var(--markdown-interelement-space-px);
}
```

Based on the above review, I think we can craft the following selector, taking in mind that the changes we want to apply are isolated in the Message Formatting section:
```
#message-formatting {
    & .preserve-message-formatting-layout {
        margin-bottom: unset;
    }
}
```
Thank you. Let me know if any further adjustments are needed.

saubhagya-patel commented:
@karlstolley I have made the proposed changes. Kindly review the PR. Thank you.

saubhagya-patel commented:
Hi @karlstolley, should I refactor this PR to address only the following issue:
- [x] The Time example output sits too high in its table row.

As for the following issue:
- [x] The thumb-up emoji is too low on “React to selected message with 👍”,

it doesn't seem to progress as expected. Could you please advise on how to proceed?

Thank you!

karlstolley commented:
Again, if you look at how `.emoji` is styled in the message area, those same styles should apply to the rendered example in the modal.

saubhagya-patel commented:
In the updated changes, I have applied the same style to the thumbs-up emoji as those used on the `.emoji` class in the message area via the `rendered-markdown.css`.

If the updated changes look good, that's great! However, if not, please take a look at the following note:

I believe the message area, as mentioned in [this comment](https://github.com/zulip/zulip/pull/31466#issuecomment-2471575455), is as follows:

<details>
<summary>Screenshot of message area</summary>

![Screenshot](https://github.com/user-attachments/assets/22b20d7b-c9da-4c1b-9ddf-526c1dec4454)

</details>

Please note that in the message area, the `:thumbs-up:` emoji doesn’t seem to align properly with the text as the `:heart:` emoji does. That’s why I considered applying `vertical-align: sub;` in the modal. I would like to hear your thoughts on these changes.
Thank you.

saubhagya-patel commented:
@karlstolley, can you please take a look at this PR once you have a moment. I would like to work on this.

 - In the current set of changes, I have applied the same styles as the `.emoji` class to the `:thumbs-up` emoji, stated in [this](https://github.com/zulip/zulip/pull/31466#issuecomment-2471575455) comment.

Thank you.

karlstolley commented:
Emoji are aligned based on their bounding box, not their drawn features. So while it might be expected to have the wrist of 👍🏻 align to the text, it's the bounding box that has to be respected.

saubhagya-patel commented:
@karlstolley, I tried fixing the layout of the thumbs-up emoji but wasn't able to find a good workaround using the bounding box alignment.

Moving forward, I have updated this PR to address only the following part of the issue:
- [x]  The Time example output sits too high in its table row.

I hope the updated changes look good to you. Please let me know if any further adjustments are needed. Thank you.

karlstolley commented:
This version of the PR focused on the `<time>` element looks to be okay, but your commit message should be rewritten to better indicate the changes here (try summarizing the more helpful long-form comment in the body of your commit).

saubhagya-patel commented:
Following the review received I have updated this PR. The following changes are made:
 - The commit message is rewritten following [this](https://github.com/zulip/zulip/pull/31466#issuecomment-2555942519) comment.
 - No visual changes are made.
 
 @karlstolley, I have made the required changes. This PR is ready for your review. Please let me know if any further adjustments are needed. Thank you.

saubhagya-patel commented:
@karlstolley, can you please take a look at this PR once you have a moment?
Thank you.

karlstolley commented on web/src/info_overlay.ts:
```diff
@@ -359,3 +360,9 @@ export function show(target: string | undefined): void {
     }
 }
+
+function process_message_formatting_layout($rendered_html: JQuery): void {
```
I wonder if this shouldn't be renamed and placed in `rendered_markdown.ts`. It's possible that this might be a useful function to have elsewhere, and it seems likelier to be overlooked in `info_overlay.ts`.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:
- The function `process_message_formatting_layout()` is now moved from `info_overlay.ts` to `rendered_markdown.ts` without being renamed as stated in [this](https://github.com/zulip/zulip/pull/31466#discussion_r1904668570) comment.

- Added `<hr>` in the elements to which the class (`preserve-message-formatting-layout`) is applied this centers the horizontal line vertically in its cell.

- **Visual changes:**
  <details>
  <summary>Applied changes to <code>---</code> markdown row.</summary>
  
  | Before | After |
  |--------|-------|
  | ![image](https://github.com/user-attachments/assets/e26ada52-a69e-4149-a5de-8a0d9a100c3a) | ![image](https://github.com/user-attachments/assets/32542df6-3813-480b-b2d3-4d49164250d6) |
  
  </details>
  
@karlstolley, this PR is ready for review. I hope the updated changes look good to you. Please let me know if any further adjustments are needed.

---
  

I wanted to state a little improvement here. Can we refactor the function `process_message_formatting_layout()` as follows:

```ts

export function add_class_to_rendered_html(
    $rendered_html: JQuery,
    css_selector: string,
    classname: string,
): void {
    const $elements = $rendered_html.find(css_selector.trim());
    if ($elements.length === 0) {
        return;
    }
    $elements.each(function () {
        $(this).addClass(classname);
    });
}

```
I would like to hear your thoughts on the function and its name.
Thank you.

saubhagya-patel commented:
@karlstolley, could you please take a look at this PR, once you have a moment?
It is quite close to completion. Thank you.

karlstolley commented:
Looks like the space has collapsed around the "Detailed message formatting documentation" line. Please look into that regression.

saubhagya-patel commented:
@karlstolley, the regression was resolved during the latest push.
Thank you.

saubhagya-patel commented:
@karlstolley, could you please take a look at this PR once you have a moment?
Thanks.

saubhagya-patel commented:
> Is there any update on this ?

Awaiting review from @karlstolley.

karlstolley commented on web/styles/informational_overlays.css:
```diff
@@ -121,3 +121,9 @@
+
+#message-formatting {
+    & .preserve-message-formatting-layout {
+        margin-bottom: unset;
```
This seems like a whole lot of JavaScript for what's ultimately a single CSS property that needs to be adjusted.

Do you think that there are other places in the UI where the `.preserve-message-formatting-layout` class would be useful?

saubhagya-patel replied:
>Do you think that there are other places in the UI where the .preserve-message-formatting-layout class would be useful?

I can't think of them at this moment, but if we want the function to be more reusable, we can refactor the function `process_message_formatting_layout()` as follows:
```ts
export function add_class_to_rendered_html(
    $rendered_html: JQuery,
    css_selector: string,
    classname: string,
): void {
    const $elements = $rendered_html.find(css_selector.trim());
    if ($elements.length === 0) {
        return;
    }
    $elements.each(function () {
        $(this).addClass(classname);
    });
}
```

Thank you.
